### PR TITLE
BUG: Improved error msg

### LINKED ISSUE #9259: Bug: Pandas indexing with at/iat fails with multi-index
Split off from: #9250

```
In [6]:

# Dataframe now:

cols = [('a_col', chr(i+65)) for i in range(5)]
cols.extend([('b_col', chr(i+65)) for i in range(5, 10)])

idx = [('a_row', chr(i+65)) for i in range(5)]
idx.extend([('b_row', chr(i+65)) for i in range(5, 10)])

df = pd.DataFrame(np.linspace(1, 100, 100).reshape(10, 10),
                  index = pd.MultiIndex.from_tuples(idx),
                  columns = pd.MultiIndex.from_tuples(cols))

In [7]:

df
Out[7]:
a_col   b_col
A   B   C   D   E   F   G   H   I   J
a_row   A   1   2   3   4   5   6   7   8   9   10
B   11  12  13  14  15  16  17  18  19  20
C   21  22  23  24  25  26  27  28  29  30
D   31  32  33  34  35  36  37  38  39  40
E   41  42  43  44  45  46  47  48  49  50
b_row   F   51  52  53  54  55  56  57  58  59  60
G   61  62  63  64  65  66  67  68  69  70
H   71  72  73  74  75  76  77  78  79  80
I   81  82  83  84  85  86  87  88  89  90
J   91  92  93  94  95  96  97  98  99  100
In [8]:

df.loc[('a_row', 'A')] # Works
df.loc[:, ('a_col', 'A')] # Works
df.ix[('a_row', 'B')] # Works
Out[8]:
a_col  A    11
       B    12
       C    13
       D    14
       E    15
b_col  F    16
       G    17
       H    18
       I    19
       J    20
Name: (a_row, B), dtype: float64
In [9]:

df.at[('a_row', 'A')] # fails
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-9-6b9462dd2fc9> in <module>()
----> 1 df.at[('a_row', 'A')] # fails

/home/federico/Python_Libraries/pandas/pandas/core/indexing.pyc in __getitem__(self, key)
   1505 
   1506         key = self._convert_key(key)
-> 1507         return self.obj.get_value(*key, takeable=self._takeable)
   1508 
   1509     def __setitem__(self, key, value):

/home/federico/Python_Libraries/pandas/pandas/core/frame.pyc in get_value(self, index, col, takeable)
   1640             return _maybe_box_datetimelike(series.values[index])
   1641 
-> 1642         series = self._get_item_cache(col)
   1643         engine = self.index._engine
   1644         return engine.get_value(series.get_values(), index)

/home/federico/Python_Libraries/pandas/pandas/core/generic.pyc in _get_item_cache(self, item)
   1066         res = cache.get(item)
   1067         if res is None:
-> 1068             values = self._data.get(item)
   1069             res = self._box_item_values(item, values)
   1070             cache[item] = res

/home/federico/Python_Libraries/pandas/pandas/core/internals.pyc in get(self, item, fastpath)
   2847 
   2848             if not isnull(item):
-> 2849                 loc = self.items.get_loc(item)
   2850             else:
   2851                 indexer = np.arange(len(self.items))[isnull(self.items)]

/home/federico/Python_Libraries/pandas/pandas/core/index.pyc in get_loc(self, key)
   4123 
   4124         if not isinstance(key, tuple):
-> 4125             loc = self._get_level_indexer(key, level=0)
   4126             return _maybe_to_slice(loc)
   4127 

/home/federico/Python_Libraries/pandas/pandas/core/index.pyc in _get_level_indexer(self, key, level)
   4359         else:
   4360 
-> 4361             loc = level_index.get_loc(key)
   4362             if level > 0 or self.lexsort_depth == 0:
   4363                 return np.array(labels == loc,dtype=bool)

/home/federico/Python_Libraries/pandas/pandas/core/index.pyc in get_loc(self, key)
   1405         loc : int if unique index, possibly slice or mask if not
   1406         """
-> 1407         return self._engine.get_loc(_values_from_object(key))
   1408 
   1409     def get_value(self, series, key):

/home/federico/Python_Libraries/pandas/pandas/index.so in pandas.index.IndexEngine.get_loc (pandas/index.c:3812)()

/home/federico/Python_Libraries/pandas/pandas/index.so in pandas.index.IndexEngine.get_loc (pandas/index.c:3692)()

/home/federico/Python_Libraries/pandas/pandas/hashtable.so in pandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:12318)()

/home/federico/Python_Libraries/pandas/pandas/hashtable.so in pandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:12269)()

KeyError: 'A'
```

It also fails on Series with multi-indexing.

### PR DESCRIPTION
**Pull request by a-y-khan**
- [x] closes #9259
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Hopefully this is a more informative message. The original bug report mentions that the index key also fails for Series but didn't report the error type, so I took my best guess at the reporter's intent in the test.

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -2022,16 +2022,25 @@ def _convert_key(self, key, is_setter: bool = False):
...
+                if not is_list_like_indexer(key):
+                    key = tuple([key])
+                else:
+                    raise ValueError("Invalid call for scalar access (getting)!")
 
-        key = self._convert_key(key)
-        return self.obj._get_value(*key, takeable=self._takeable)
+            key = self._convert_key(key)
+            return self.obj._get_value(*key, takeable=self._takeable)
+        except KeyError as err:
```
Comment (jreback) [inline pandas/core/indexing.py]: this try/except is pretty wide, you likely just need to capture line 2035 only.

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -2022,16 +2022,25 @@ def _convert_key(self, key, is_setter: bool = False):
...
+            if isinstance(self.obj.index, ABCMultiIndex):
+                raise KeyError(
+                    f"Detected KeyError {err}, indexing with {key} "
+                    "failing for MultiIndex"
+                ) from err
+            else:
```
Comment (jreback) [inline pandas/core/indexing.py]: can be just raise

File: pandas/tests/indexes/multi/test_get_set.py [hunk 3]
```diff
@@ -413,3 +413,23 @@ def test_set_levels_with_iterable():
         [expected_sizes, colors], names=["size", "color"]
     )
     tm.assert_index_equal(result, expected)
+
+
+def test_at_indexing_fails_multiindex():
+    # GH9259
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_get_set.py]: you don't need this complicated of a MI, just make a simple one.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by a-y-khan** [comment]
I made some changes. Please let me know if this is what you had mind. Thanks!

File: pandas/tests/indexes/multi/test_get_set.py [hunk 4]
```diff
@@ -329,3 +329,21 @@ def test_set_levels_with_iterable():
...
+    cols = [("a_col", chr(i + 65)) for i in range(2)]
+    idx = [("a_row", chr(i + 65)) for i in range(2)]
+    df = pd.DataFrame(
+        np.linspace(1, 4, 4).reshape(2, 2),
+        index=pd.MultiIndex.from_tuples(idx),
+        columns=pd.MultiIndex.from_tuples(cols),
+    )
+    s = pd.Series(np.linspace(1, 2, 2), index=pd.MultiIndex.from_tuples(idx))
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_get_set.py:343]: nitpick: can you call this `ser`?  1-letter variable names are a hassle to grep for

Comment (jbrockmendel) [reply pandas/tests/indexes/multi/test_get_set.py:343]: could also just define it `ser = df[("a_col", "A")]`

File: pandas/tests/indexes/multi/test_get_set.py [hunk 5]
```diff
@@ -329,3 +329,21 @@ def test_set_levels_with_iterable():
...
+
+    with pytest.raises(KeyError, match=r".+? indexing with .+? failing for MultiIndex"):
+        df.at["a_row", "A"]
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_get_set.py:346]: is there a similar problem for `at.__setitem__`?

File: pandas/core/indexing.py [hunk 6]
```diff
@@ -2032,7 +2032,16 @@ def __getitem__(self, key):
...
 
         key = self._convert_key(key)
-        return self.obj._get_value(*key, takeable=self._takeable)
+        try:
+            return self.obj._get_value(*key, takeable=self._takeable)
+        except KeyError as err:
+            if isinstance(self.obj.index, ABCMultiIndex):
+                raise KeyError(
+                    f"Detected KeyError {err}, indexing with {key} "
+                    "failing for MultiIndex"
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:2041]: doing this on master, the DataFrame case is giving a reasonable message, while the Series case is unhelpful.  Maybe condition this on `self.ndim == 1` and add a suggestion "pass a tuple"

File: pandas/tests/indexes/multi/test_get_set.py (hunk 4 above)
Comment (a-y-khan) [reply pandas/tests/indexes/multi/test_get_set.py]: Done!

File: pandas/tests/indexes/multi/test_get_set.py (hunk 5 above)
Comment (a-y-khan) [reply pandas/tests/indexes/multi/test_get_set.py:346]: Multiindex with at.__setitem__ doesn't raise an exception and sets items at level 0:

```
In [1]: import pandas as pd 
   ...: import numpy as np 
   ...:  
   ...: cols = [("a_col", chr(i + 65)) for i in range(2)] 
   ...: idx = [("a_row", chr(i + 65)) for i in range(2)] 
   ...: df = pd.DataFrame( 
   ...:     np.linspace(1, 4, 4).reshape(2, 2), 
   ...:     index=pd.MultiIndex.from_tuples(idx), 
   ...:     columns=pd.MultiIndex.from_tuples(cols), 
   ...: ) 
   ...: display(df)
                                                                                                           
        a_col     
            A    B
a_row A   1.0  2.0
      B   3.0  4.0

In [2]: df.at["a_row", "A"] = 888.0 
   ...: display(df)
                                                                                                           
         a_col       
             A      B
a_row A  888.0  888.0
      B    3.0    4.0

In [3]: df.at["a_row", "A"] = [888.0, 999.0] 
   ...: display(df)
                                                                                                           
         a_col       
             A      B
a_row A  888.0  999.0
      B    3.0    4.0
```

File: pandas/core/indexing.py (hunk 6 above)
Comment (a-y-khan) [reply pandas/core/indexing.py:2041]: It the Series case goes down a different code path where an exception is raised in `pandas/utils/_validators.py`. Would adding an except for TypeError with a more informative error message work?

```
In [3]: ser.at["a_row", "A"]                                                                                  
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-2aec4c8078ba> in <module>
----> 1 ser.at["a_row", "A"]

/workspaces/pandas-aykhan/pandas/core/indexing.py in __getitem__(self, key)
   2090             return self.obj.loc[key]
   2091 
-> 2092         return super().__getitem__(key)
   2093 
   2094     def __setitem__(self, key, value):

/workspaces/pandas-aykhan/pandas/core/indexing.py in __getitem__(self, key)
   2035         try:
   2036             # import pdb; pdb.set_trace()
-> 2037             return self.obj._get_value(*key, takeable=self._takeable)
   2038         except KeyError as err:
   2039             if isinstance(self.obj.index, ABCMultiIndex):

TypeError: _get_value() got multiple values for argument 'takeable'
```

File: pandas/tests/indexes/multi/test_get_set.py [hunk 7]
```diff
@@ -329,3 +329,21 @@ def test_set_levels_with_iterable():
...
+        index=pd.MultiIndex.from_tuples(idx),
+        columns=pd.MultiIndex.from_tuples(cols),
+    )
+    ser = df[("a_col", "A")]
+
+    with pytest.raises(KeyError, match=r".+? indexing with .+? failing for MultiIndex"):
+        df.at["a_row", "A"]
+
+    with pytest.raises(TypeError, match=r".+? got multiple values for argument .+?"):
+        ser.at["a_row", "A"]
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_get_set.py:349]: this looks like a bug.  `ser.at["a_row", "A"]` should return 1.0, same as `ser.loc["a_row", "A"]`

Comment (rhshadrach) [reply pandas/tests/indexes/multi/test_get_set.py:349]: I tested this on my PR #32520, it returns 1.0.

Comment (a-y-khan) [reply pandas/tests/indexes/multi/test_get_set.py:349]: I can remove the series from the test. @rhshadrach is correct that the original context was DataFrames, and improving the error message should clarify why the KeyError is being raised.

**Comment by jreback** [comment]
I am not sure this is needed any longer as we merged #32520, but pls @a-y-khan if you can test and see if the original issue now works, or we still need something.

**Comment by jreback** [comment]
closing as stale, if you'd like to update as above, pls ping / open a new PR.

**Pull request closed without merging**